### ISSUE TITLE
Versioning

### ISSUE THREAD

**Issue by mateogianolio** [mateogianolio/vectorious#7]
For the purpose of structure and consistency, we should decide how to version releases. Up until now, I have just incremented the version number when changing something: ```1.0.0 => 1.0.1 => ... => 1.0.9 => 1.1.0```.

I'm suggesting version ```2.0.0``` as the point where we've converted everything to typed arrays and successfully implemented C/C++ bindings to the extent that we can call it a stable library suitable for use in production environments.

**Comment by metabench** [comment]
After 1.0.9 could come 1.0.10.

Your suggestion looks good, but I have another suggestion. I think an intermediate point between what we have at the moment and your suggested `2.0.0` is a version that's using typed arrays and more optimized JavaScript, but still does not use C++. That would be `2.0.0`, with the C++ version likely being `3.0.0`.

Another thing to consider is some kind of bleeding-edge fork or branch. I'm thinking that SIMD would provide very large performance improvements, and be ready for production without much effort, but only be ready on some particular platforms.

**Comment by metabench** [comment]
For the meantime, where do we stand with version numbers, and deploying recent changes to npm?

**Comment by mateogianolio** [comment]
I added two milestones (2.0.0 and 3.0.0) to keep better track of issues.

Maybe we could have some sort of system, like ```x.y.z```, where ```x``` is breaking API changes, ```y``` is major optimizations or API additions and ```z``` is small bug fixes or minor improvements.

I seriously don't know, versions confuse me. Maybe you have a better system? :)

**Comment by metabench** [comment]
What you have described is pretty much the the 'semantic versioning' or 'semver' system. I think that putting breaking API changes into the major version number makes the most sense.

Also, I think large architectural changes would merit another version number. Arguably Typed Arrays would change the API or it's maybe compatibility, so I think that changing to Typed Arrays would warrant a major version too.

Basically I think `2.0.0` should contain the recent fixes and optimizations, and be released relatively soon.

**Comment by mateogianolio** [comment]
Ok, so you're suggesting we introduce the recent optimizations in ```2.0.0``` and let ```3.0.0``` be typed arrays and ```4.0.0``` C/C++ bindings?

**Comment by metabench** [comment]
Yes, I think that's a good sequence to plan for. However, if for some reason we think it's worth introducing an API change in the meantime let's not be too rigid about sticking to that versioning. I kind-of doubt that will happen before `3.0.0` but I'm mentioning it anyway.

The C++ bindings seem like quite a big challenge. I don't know how long it will take. However, version `4.0.0` having the C++ bindings looks like a sensible goal at this point.

Let's also make Version `2.0.0` fully / much better documented before it gets released.

**Comment by mateogianolio** [comment]
Agreed.

**Comment by mateogianolio** [comment]
We now have milestones for versions ```2.0.0```, ```3.0.0``` and ```4.0.0``` and we have a pretty good understanding of how to work, so I consider this issue resolved at least for now.

**Comment by metabench** [comment]
I was getting annoyed by the milestones not showing in order when displayed, so I set some due dates, giving far more than enough time. Then I think we should bring the due dates forward once we know what a release schedule actually is.
